module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+
+		accepted := e.fetchQueue.Push(request)
+		if !accepted {
+			// Silently drop requests if the queue is full. They will be backfilled when the next
+			// finalized block is processed. This takes advantage of the fact that seals are
+			// guranteed by hotstuff to be contiguous.
+			logger.Warn().
+				Str("sealed_block_id", request.blockID.String()).
+				Uint64("sealed_block_height", request.height).
+				Msg("fetch request queue is full")
+			e.notificationState.Sealed(highestSealed)
+			return nil
+		}
+
+		logger.Debug().Msgf("queue has %d requests", e.fetchQueue.Len())
+
+		highestSealed = request.height
+		e.fetchNotifier.Notify()
+	}
+
+	e.notificationState.Sealed(highestSealed)
+	return nil
+}
+
+func (e *executionDataRequesterImpl) missingBefore(current uint64, end uint64) ([]fetchRequest, error) {
+	missedHeights := []fetchRequest{}
+
+	for current < end {
+		e.log.Debug().Msgf("adding missed sealed height: %d, latest: %d", current, end)
+		missedBlock, err := e.blocks.ByHeight(current)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get block for height %d: %w", current, err)
+		}
+
+		missedHeights = append(missedHeights, fetchRequest{
+			blockID: missedBlock.ID(),
+			height:  current,
+		})
+
+		current++
+	}
+
+	return missedHeights, nil
+}
+
+// Fetch Worker Methods
+
+func (e *executionDataRequesterImpl) fetchRequestProcessingLoop(ctx irrecoverable.SignalerContext) {
+
+	e.consumerMu.Lock()
+	workerID := e.workerCount
+	e.workerCount++
+	e.consumerMu.Unlock()
+
+	logger := e.log.With().Int("worker_id", workerID).Logger()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-e.stop:
+			return
+		case <-e.fetchNotifier.Channel():
+			logger.Debug().Msgf("fetching available requests")
+			e.processAvailableFetchRequests(ctx)
+			logger.Debug().Msgf("done fetching requests")
+		}
+	}
+}
+
+func (e *executionDataRequesterImpl) processAvailableFetchRequests(ctx irrecoverable.SignalerContext) {
+
+	e.log.Debug().Msg("processing fetch requests")
+	for {
+		// retries get priority since the seal height is most likely lower
+		item, ok := e.fetchRetryQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// Check for missing backfills
+
+		item, ok = e.fetchQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// return when queue is empty
+		return
+	}
+}
+
+func (e *executionDataRequesterImpl) processFetchRequest(ctx irrecoverable.SignalerContext, request fetchRequest) error {
+	logger := e.log.With().Str("block_id", request.blockID.String()).Logger()
+
+	logger.Debug().Msgf("processing fetch request for block %d", request.height)
+
+	executionData, err := e.byBlockID(ctx, request.blockID)
+
+	logger.Debug().Msgf("done processing fetch request for block %d", request.height)
+
+	if errors.Is(err, &state_synchronization.MalformedDataError{}) || errors.Is(err, &state_synchronization.BlobSizeLimitExceededError{}) {
```

Comment:
Can we make sure to test this? I think you actually need `errors.As` here, don't think `errors.Is` will work because that checks for exact match, which you won't get.

Reply:
fixed in the lastest version